<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-License-Identifier: BSD-3-Clause
SPDX-FileCopyrightText: Czech Technical University in Prague
-->

<launch>
    <arg name="device" default="/dev/septentrio_usb2" />
    <arg name="config" default="$(find-pkg-share cras_septentrio_gnss_driver)/config/septentrio_ros2.yaml" />
    <arg name="enable_raw_output" default="true" />

    <arg name="ntrip_caster" default="ntrip.pecny.cz" />
    <arg name="ntrip_caster_port" default="2101" />
    <arg name="ntrip_mountpoint" default="GOP60" />

    <arg name="namespace" default="gnss/septentrio" />
    <arg name="component_container" default="septentrio_container" />
    <arg name="run_component_container" default="true" />

    <arg name="min_position_error" default="0.0" />
    <arg name="min_altitude_error" default="0.0" />
    <arg name="min_heading_error_deg" default="0.0" />
    <arg name="publish_invalid_fix" default="true" />
    <arg name="publish_invalid_heading" default="false" />
    <arg name="valid_position_error_threshold" default="1.0" />
    <arg name="valid_heading_error_threshold_deg" default="10.0" />

    <arg name="debug" default="false" />
    <arg name="output" default="screen" />
    <arg name="respawn" default="true" />

    <!-- End of arguments -->

    <let name="device_name" value="serial:$(var device)" unless="$(eval '\':\' in \'$(var device)\'')" />
    <let name="device_name" value="$(var device)" if="$(eval '\':\' in \'$(var device)\'')" />

    <let name="custom_commands"
         value="$(find-pkg-share cras_septentrio_gnss_driver)/config/enable_raw_output.sbf_commands"
         if="$(var enable_raw_output)"/>
    <let name="custom_commands" value=""
         unless="$(var enable_raw_output)"/>

    <group>
        <push_ros_namespace namespace="$(var namespace)" />
        <node pkg="rclcpp_components" exec="component_container_mt" name="$(var component_container)"
              output="$(var output)" respawn="$(var respawn)" if="$(var run_component_container)">
        </node>
        <load_composable_node target="$(var namespace)/$(var component_container)">
            <composable_node name="driver" pkg="septentrio_gnss_driver" plugin="rosaic_node::ROSaicNode" namespace="raw">
                <param from="$(var config)" />

                <param name="device" value="$(var device_name)" />
                <param name="activate_debug_log" value="$(var debug)" />
                <param name="custom_commands_file" value="$(var custom_commands)" />

                <param name="rtk_settings.ntrip_1.caster" value="$(var ntrip_caster)" />
                <param name="rtk_settings.ntrip_1.caster_port" value="$(var ntrip_caster_port)" />
                <param name="rtk_settings.ntrip_1.mountpoint" value="$(var ntrip_mountpoint)" />

                <remap from="navsatfix" to="fix" />
                <remap from="gpsfix" to="fix_detail" />

                <extra_arg name="use_intra_process_comms" value="true" />
            </composable_node>

            <composable_node name="septentrio_process" pkg="cras_septentrio_gnss_driver" plugin="cras::SeptentrioProcess">
                <param name="min_position_error" value="$(var min_position_error)" />
                <param name="min_altitude_error" value="$(var min_altitude_error)" />
                <param name="min_heading_error_deg" value="$(var min_heading_error_deg)" />
                <param name="publish_invalid_fix" value="$(var publish_invalid_fix)" />
                <param name="publish_invalid_heading" value="$(var publish_invalid_heading)" />
		<param name="valid_position_error_threshold" value="$(var valid_position_error_threshold)" />
                <param name="valid_heading_error_threshold_deg" value="$(var valid_heading_error_threshold_deg)" />

                <extra_arg name="use_intra_process_comms" value="true" />
            </composable_node>
        </load_composable_node>
    </group>
</launch>
